### PR TITLE
#349 Remove ee_models_library.csv

### LINKED ISSUE #349: Remove ee_models_library.csv
## Context

Once #335 is solved, it is not necessary to have the file `ee_models_library.csv` any more.

Indeed, such file was needed to know which NER model to use in order to extract a given entity type, but we have decided (see also #342) that from now on for each entity type there will be a dedicated NER model stored under ` data_and_models/models/ner_er/model-<entity_type_name>/`.

## Actions
- [x] remove `ee_models_library.csv`
- [x] remove `load_ee_models_library()`
- [x] modify all functions that relied on `ee_models_library.csv` or `load_ee_models_library()`

### PR DESCRIPTION
**Pull request by EmilieDel**
Fixes #349.

## Description

Once #335 is solved, it is not necessary to have the file `ee_models_library.csv` any more.

**Now, each entity model is detecting one entity type (and has a name `model-<entity-type-name>`)**

What was done during in this PR ?
- Removed `bluesearch.utils.load_ee_models_library` and `ee_models_library.csv` and corresponding tests
- Added `get_available_spacy_models` in the `utils` package and corresponding tests
- Modified `CreateMiningCache` class (+ all `linked` files: `tests`, `entrypoints`)
- Modified `MiningServer` class (+ all `linked` files: `tests`, `entrypoints`)
- Modified `bluesearch.utils.retrieve_mining_cache` with the new logic and corresponding tests.

## Checklist

- [X] This PR refers to an issue from the [issue tracker](https://github.com/BlueBrain/Search/issues).
  (if it is not the case, please create an issue first).
- [X] Unit tests added.
  (if needed)
- [X] Documentation and `whatsnew.rst` updated.
  (if needed)
- [X] `setup.py` and `requirements.txt` updated with new dependencies.
  (if needed)
- [X] Type annotations added.
  (if a function is added or modified)
- [X] All CI tests pass. 

## Mining Cache Creation
| Entity Type | Before | After (14.06) | Change % | 
| -- | -- | -- | -- |
| Cell Compartment | 852'704 | 691'613 | -18.9 % |
| Cell Type | 1'902'729 | 1'370'991 | -27.95 % |
| Chemical | 3'439'474 | 1'725'962 | -49.82 % |
| Disease | 7'246'744 | 5'035'939 | -30.51 % |
| Drug | 1'558'625 | 1'014'880 | -34.89 % |
| Organ | 3'727'866 | 3'165'154 | -15.10 % |
| Organism | 8'230'190 | 6'647'255 | -19.24 % |
| Pathway | 1'777'978 | 1'299'866 | -26.80 % |
| Protein | 8'386'713 | 3'744'740 | -55.35 % |
| -- | -- | -- | -- | 
| Total | 37'123'023 | 24'696'400 | -33.48 % |

## Potential future TO DOs

Those are points found out during this PR that could be worth addressing in the future (maybe we can open issues):
- The `MiningServer` class is expecting `model_libs` with three keys `ee`, `ae`, `re`. Maybe we can think about simplifying this ?
- Are the `SPECS` from `bluesearch/mining/pipeline.py` something still useful ? If yes, should it still be the responsibility of `run_pipeline` method or the one of the `MiningServer` when sending back the response ?
- The new assumption is "One entity type per model", however we still use a default spacy model in our tests with several entity types. We could try to switch for another model that reflects better the reality.
- `MiningServer` has a lot of copy/paste in the logic i.e. `pipeline_text`, `pipeline_database`, it could be nice to think about a refactoring.
- Clean `data_and_models` part and remove the implicit `ee_models_library` logic there (there should not have any mapping anymore). 
- [[From @pafonta](https://github.com/BlueBrain/Search/pull/366#pullrequestreview-664626921)] Simplify the handling of NER models by introducing a class representing the information of a NER model (i.e. recognized entity type name, full pathlib.Path, ...).

### CODE REVIEW

**Review by pafonta — COMMENTED** [review]
Thank you for making the codebase simpler!

I have one main question and one suggestion:

## Question

It seems there is a leftover in `src/bluesearch/mining/entity.py`. Indeed, the method `global2model_patterns` seems to still use a `ee_models_library`-style logic.

**Shouldn't it be changed too?**

## Suggestion

For this PR `or` as a potential future TODO, I would add the following:

`Simplify the handling of NER models by introducing a class representing the information of a NER model` (i.e. recognized entity type name, full pathlib.Path, ...).

➡️ It would `remove` a lot of `complexity` of the handling of the dictionary [etype, model_path] and the passing of the etype argument. Besides, the interface / API of our methods and functions would become more harmonized, which would increase modularity.

**Comment by EmilieDel** [comment]
Thanks a lot for your feedback, @pafonta and great finding! 
I tried to address your comment in the [last commit of this PR](https://github.com/BlueBrain/Search/pull/366/commits/5bfa977e7c8253b33dd562627e0bb8b69371eea0). Do not hesitate to have a look. 
As for your suggestion, I added it in the list of `Potential future TO DOs`. Maybe we can start creating tickets for some of them! What do you think ?

**Comment by pafonta** [comment]
Hello @EmilieDel,

> I tried to address your comment in the [last commit of this PR](https://github.com/BlueBrain/Search/pull/366/commits/5bfa977e7c8253b33dd562627e0bb8b69371eea0). Do not hesitate to have a look.

Thank you! I did. LGTM 👍 

> As for your suggestion, I added it in the list of `Potential future TO DOs`. Maybe we can start creating tickets for some of them! What do you think ?

I think that's a good idea!

**Review by pafonta — APPROVED** [review]

File: tests/conftest.py [hunk 1]
```diff
@@ -310,11 +309,33 @@ def metadata_path():
...
+    models_dir = data_dir / "models" / "ner_er"
+    models_dir.mkdir(parents=True)
+
+    for etype in entity_types:
+        etype_model_dir = models_dir / f"model-{etype.lower()}"
+        etype_model_dir.mkdir()
+        meta_path = etype_model_dir / "meta.json"
+        with open(meta_path, "w") as f:
+            f.write(
+                json.dumps(
```
Comment (Stannislav) [inline tests/conftest.py:328]: I think this can be simplified to `json.dump(data, fp)`

File: tests/test_utils.py [hunk 2]
```diff
@@ -345,26 +352,56 @@ def test_load_save_jsonl(tmpdir):
...
+
+    # Missing meta.json file
+    good_model_dir = pathlib.Path(tmpdir) / "models" / "ner_er" / "model-organism"
+    good_model_dir.mkdir(parents=True)
+    assert check_entity_type_consistency(good_model_dir) is False
+
+    # meta.json structure
+    meta_path = good_model_dir / "meta.json"
+    with open(meta_path, "w") as f:
+        f.write(json.dumps(metadata))
```
Comment (Stannislav) [inline tests/test_utils.py:380]: Same as above, one can write `json.dump(metadata, f)`

File: tests/test_utils.py [hunk 3]
```diff
@@ -345,26 +352,56 @@ def test_load_save_jsonl(tmpdir):
...
+    assert results == expected_results
+
+
+@pytest.mark.filterwarnings()
+def test_get_available_spacy_models_warning(tmpdir):
+    """Test that available spacy models warnings."""
+    wrong_path = pathlib.Path(tmpdir) / "wrongpath"
+    wrong_model = wrong_path / "models" / "ner_er" / "wrongmodel"
+    wrong_model.mkdir(parents=True)
+    with pytest.warns(None):
```
Comment (Stannislav) [inline tests/test_utils.py:403]: Isn't `None` the default argument here anyway? The type hint for this arg is `Optional[...]`.

So with `None`, does it test for all warning types? If we know which warning will be raised, maybe we can specify the exact warning type here?

**Review by Stannislav — APPROVED** [review]
Thanks :)

File: tests/conftest.py (hunk 1 above)
Comment (EmilieDel) [reply tests/conftest.py]: Done in 7f5c0d6.

File: tests/test_utils.py (hunk 2 above)
Comment (EmilieDel) [reply tests/test_utils.py]: Done in 7f5c0d6.

File: tests/test_utils.py (hunk 3 above)
Comment (EmilieDel) [reply tests/test_utils.py:403]: It is to catch the `warning` if something is not consistent in the folder of the spacy models (wrong path, wrong model name, etc. ). We did not specified any type of `warnings` for now. Do you think we should do it ? If yes, do you have a preference ?

Comment (Stannislav) [reply tests/test_utils.py:403]: I don't remember what we do in the code, but by default if you do `warnings.warn(message)` it creates a `UserWarning`. So I was wondering what pytest checks for if you write `with pytest.warns(None)`? That a warning of any type is issued?

Comment (EmilieDel) [reply tests/test_utils.py:403]: I would say so! But that is a good point, I can flag `UserWarning`.

**Comment by EmilieDel** [comment]
Hello everyone ☀️ 
I instantiated a `mining_server` on DGX at port `9852` if you want to try it by yourself!

**Comment by FrancescoCasalegno** [comment]
## To Do's before merging
- [ ] have 2 people to try the new server
- [ ] create issue to ensure that huggingface transformer can handle (= change its config) long sentences
- [ ] create issue to improve filtering of sentences to be added to the database (filter out latex content, sentences that are too long/too short, ...) -- mark as `bad` or just don't add to database

**Comment by pafonta** [comment]
Hello @EmilieDel,

> I instantiated a `mining_server` on DGX at port `9852` if you want to try it by yourself!

The mining widget worked for me.

However, I am surprised to see `This server is using the database: cord19_v47`.

**Haven't we moved to CORD-19 `v65` in December, 2020?**

**Comment by FrancescoCasalegno** [comment]
@EmilieDel feel free to merge!

Please just remember to take care of all the consequences of this merge — e.g. killing server on `8892`, maybe re-launching servers, ...

**Pull request merged**